### PR TITLE
Balder/smarter same element iteration

### PR DESCRIPTION
**Pull request by baldersheim**
@havardpe PR

### CODE REVIEW

File: searchlib/src/vespa/searchlib/attribute/CMakeLists.txt [hunk 1]
```diff
@@ -34,7 +34,7 @@ vespa_add_library(searchlib_attribute OBJECT
     defines.cpp
     diversity.cpp
     dociditerator.cpp
-    elementiterator.cpp
+        searchcontextelementiterator.cpp
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/attribute/CMakeLists.txt:37]: strange indent

File: searchlib/src/vespa/searchlib/attribute/searchcontextelementiterator.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    }
+}
+void
+SearchContextElementIterator::mergeElementIds(uint32_t docId, std::vector<uint32_t> & elementIds) {
+    size_t toKeep(0);
+    int32_t id(-1);
+    int32_t weight(0);
+    for (int32_t candidate : elementIds) {
+        if (candidate >= id) {
+            id = _searchContext.find(docId, candidate, weight);
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/attribute/searchcontextelementiterator.cpp:25]: possible re-write: avoid calling find multiple times for same candidate
possible re-write: terminate loop when attribute matching is exhausted.

File: searchlib/src/vespa/searchlib/queryeval/elementiterator.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    } else {
+        self.visitNull(name);
+    }
+}
+
+void visit(vespalib::ObjectVisitor &self, const vespalib::string &name,
+           const search::queryeval::ElementIterator &obj)
+{
+    visit(self, name, &obj);
+}
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/queryeval/elementiterator.cpp:71]: missing newline

File: searchlib/src/vespa/searchlib/queryeval/elementiterator.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+void
+ElementIteratorWrapper::mergeElementIds(uint32_t docId, std::vector<uint32_t> & elementIds) {
+    _search->unpack(docId);
+    size_t toKeep(0);
+    int32_t id(-1);
+    auto it = _tfmd.begin();
+    for (int32_t candidate : elementIds) {
+        if (candidate >= id) {
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/queryeval/elementiterator.cpp:39]: possible re-write: avoid seeking for the same candidate multiple times

**Review by havardpe — APPROVED** [review]
I like this a lot 👍

File: searchlib/src/vespa/searchlib/queryeval/elementiterator.cpp (hunk 3 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/queryeval/elementiterator.cpp:71]: Fixed

File: searchlib/src/vespa/searchlib/attribute/CMakeLists.txt (hunk 1 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/CMakeLists.txt:37]: Fixed

File: searchlib/src/vespa/searchlib/queryeval/elementiterator.cpp (hunk 4 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/queryeval/elementiterator.cpp]: Fixed

File: searchlib/src/vespa/searchlib/attribute/searchcontextelementiterator.cpp (hunk 2 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/searchcontextelementiterator.cpp:25]: Fixed

**Pull request merged**